From 95b677c847db9efbffd7c133f14225a3dca7cd7f Mon Sep 17 00:00:00 2001
From: Marius Vlad <marius.vlad@collabora.com>
Date: Tue, 30 Mar 2021 19:26:40 +0300
Forwarded: not-needed
Subject: [PATCH 14/14] kiosk-shell: Avoid input focus and xdg-shell activation

For views that make into the pop-up layer (notifications), we do not
activate any focus on those surfaces such that anything that is lower
in the stack order keeps the input focus (this way bundles continue
to receive inputs).

Signed-off-by: Marius Vlad <marius.vlad@collabora.com>
---
 kiosk-shell/kiosk-shell.c | 32 +++++++++++++++++++-------------
 1 file changed, 19 insertions(+), 13 deletions(-)

diff --git a/kiosk-shell/kiosk-shell.c b/kiosk-shell/kiosk-shell.c
index 1baca36e..cfc454dc 100644
--- a/kiosk-shell/kiosk-shell.c
+++ b/kiosk-shell/kiosk-shell.c
@@ -833,20 +833,26 @@ desktop_surface_committed(struct weston_desktop_surface *desktop_surface,
 
 		wl_list_for_each(seat, &shsurf->shell->compositor->seat_list, link) {
 			struct kiosk_shell_seat *kiosk_seat = get_kiosk_shell_seat(seat);
-			weston_view_activate(shsurf->view, seat, 0);
 
-			if (kiosk_seat->focused_surface) {
-				struct kiosk_shell_surface *current_focus =
-					get_kiosk_shell_surface(kiosk_seat->focused_surface);
-				if (current_focus)
-					kiosk_shell_surface_deactivate(current_focus);
-			}
-
-			kiosk_seat->focused_surface = surface;
-
-			if (kiosk_seat->focused_surface) {
-				if (shsurf)
-					kiosk_shell_surface_activate(shsurf);
+			/* skips any input focus or xdg-shell activate request
+			 * for surfaces placed in the pop-up layer (i.e.,
+			 * notifications) */
+			if (layer_list != &shsurf->shell->popup_layer.view_list) {
+				weston_view_activate(shsurf->view, seat, 0);
+
+				if (kiosk_seat->focused_surface) {
+					struct kiosk_shell_surface *current_focus =
+						get_kiosk_shell_surface(kiosk_seat->focused_surface);
+					if (current_focus)
+						kiosk_shell_surface_deactivate(current_focus);
+				}
+
+				kiosk_seat->focused_surface = surface;
+
+				if (kiosk_seat->focused_surface) {
+					if (shsurf)
+						kiosk_shell_surface_activate(shsurf);
+				}
 			}
 		}
 	}
-- 
2.29.2

